graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/DataTypeMapping.java

```diff
@@ -15,53 +15,118 @@
  */
 package io.stargate.graphql.schema.fetchers.dml;
 
+import com.datastax.oss.driver.api.core.CqlIdentifier;
+import com.datastax.oss.driver.api.core.data.UdtValue;
+import com.datastax.oss.driver.api.core.type.codec.TypeCodec;
+import com.datastax.oss.driver.api.querybuilder.QueryBuilder;
+import com.datastax.oss.driver.api.querybuilder.term.Term;
 import io.stargate.db.datastore.Row;
 import io.stargate.db.schema.Column;
 import io.stargate.db.schema.Table;
+import io.stargate.db.schema.UserDefinedType;
 import io.stargate.graphql.schema.NameMapping;
 import java.util.Collection;
 import java.util.HashMap;
-import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
 
 /** Provides the logic for adapting values from graphql to DB and vice versa. */
 class DataTypeMapping {
-  @SuppressWarnings("unchecked")
-  static Object toDbValue(Column.ColumnType type, Object value) {
-    if (value == null || !type.isCollection()) {
-      return value;
-    }
 
-    if (type.rawType() == Column.Type.List) {
-      if (!type.parameters().get(0).isCollection()) {
-        return value;
-      }
+  /**
+   * Converts a value coming from the GraphQL runtime into a term that can be used with {@link
+   * QueryBuilder}.
+   */
+  static Term toCqlTerm(Column.ColumnType type, Object value, NameMapping nameMapping) {
+    StringBuilder raw = new StringBuilder();
+    format(type, value, nameMapping, raw);
+    return QueryBuilder.raw(raw.toString());
```

Comment:
I switched to a different approach here, the conversion was wasteful for complex objects: we received a GraphQL object, converted it to the driver equivalent, only to immediately turn it into a CQL literal. This was particularly annoying with UDTs because I would have had to reconstruct a `UdtValue`. It also happened with collections on a lesser scale.

Now the code tries to format directly into a string.

The double conversion still happens for custom scalars, but it gives us a layer of validation for types that are represented as GraphQL strings.

Reply:
Can we avoid formatting to strings and use the java types (w/ query markers) in the query builder instead?

tbh, I wasn't aware `literal()` was representing them as strings and back.

Reply:
I've filed #273 to consider switching to query markers and prepared statements.

Reply:
👍  I was thinking about that too. We'll have to figure out which types the persistence API's `PreparedStatement` expects, but it would be cleaner indeed.
And maybe we can come up with a caching scheme to avoid rebuilding queries on every call.